MGMT-8186 SRO - Hub and spoke cluster support (#46613)
openshift/openshift-docs: modules/psap-special-resource-operator-using-CRD.adoc

// Module included in the following assemblies:
//
// * hardware_enablement/psap-special-resource-operator.adoc

:_content-type: PROCEDURE

> [id="deploy-simple-kmod-using-CRD"]

Author: Done

> = Building and running the simple-kmod SpecialResource for a hub and spoke topology

Reviewer: [Throughout] "hub-and-spoke" is a compound adjective
```suggestion
= Building and running the simple-kmod SpecialResource for a hub-and-spoke topology
```

Author: changed all instances

> You can use the Special Resource Operator (SRO) on a hub and spoke deployment in which Advanced Cluster Management connects a hub cluster to one or more managed clusters.

Author: Done

> This example shows how the Special Resource Operator builds driver containers in the hub, watching hub cluster resources to identify OCP versions for the helm charts that it uses to create resources which it delivers to spokes.

Reviewer 2: I'm sorry, I don’t understand this description. Did you mean: 
```suggestion
This example shows how the Special Resource Operator builds driver containers in the hub. SRO watches the hub cluster resources to identify OCP versions for the helm charts that it uses to create resources which it delivers to spokes.
```

Author: Done

> This example procedure shows how the Special Resource Operator builds driver containers in the hub. SRO watches hub cluster resources to identify OCP versions for the helm charts that it uses to create resources which it delivers to spokes.

Reviewer: - Is this line break meant to create a new paragraph? If so, needs one more for a blank line.
- SRO is already defined above.
- SRO needs an article.
- Can't use "OCP" in docs.
```suggestion
This example procedure shows how the SRO builds driver containers in the hub. The SRO watches hub cluster resources to identify {product-title} versions for the helm charts that it uses to create resources which it delivers to spokes.
```

Author: Done

> .Prerequisites

Reviewer: Exemplary prereq section :tada:

* You have a running {product-title} cluster.
* You installed the OpenShift CLI (`oc`).
* You are logged into the OpenShift CLI as a user with `cluster-admin` privileges.
* You installed the SRO.
* You installed the Helm CLI (`helm`).

> * You installed the Advanced Cluster Management (ACM) operator.

Author: Done

* You configured a container registry.

.Procedure

> . Create a templates directory and change into it.
> +
> [source,terminal]
> ----
> $ mkdir -p charts/acm-simple-kmod-0.0.1/templates
> ----
> +
> [source,terminal]
> ----
> $ cd charts/acm-simple-kmod-0.0.1/templates
> ----

Reviewer: Commands should not be combined, and also need an intro phrase. I know it seems excessive for simple stuff like this, sorry about that :bow: 

```suggestion
. Create a `templates` directory by running the following command:
+
[source,terminal]
----
$ mkdir -p charts/acm-simple-kmod-0.0.1/templates
----

. Change to the `templates` directory by running the following command:
+
[source,terminal]
----
$ cd charts/acm-simple-kmod-0.0.1/templates
----
```

> . `BuildConfig`, `Policy`, and `PlacementRule` resources.

Author: Done

> . Create the buildconfig, policy, and placement.

Author: Done

                                                command: ["modprobe", "-r", "-a" , "simple-kmod", "simple-procfs-kmod"]
                                      securityContext:
                                        privileged: true
----
.. Save this YAML template for the placement of policies in the `templates` directory as `0003-policy.yaml`.
+
[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
    name: {{.Values.specialResourceModule.metadata.name}}-placement
spec:
    clusterConditions:
    - status: "True"
      type: ManagedClusterConditionAvailable
    clusterSelector:
      matchExpressions:
      - key: name
        operator: NotIn
        values:
        - local-cluster
---
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
    name: {{.Values.specialResourceModule.metadata.name}}-binding
placementRef:
    apiGroup: apps.open-cluster-management.io
    kind: PlacementRule
    name: {{.Values.specialResourceModule.metadata.name}}-placement
subjects:
    - apiGroup: policy.open-cluster-management.io
      kind: Policy
      name: policy-{{.Values.specialResourceModule.metadata.name}}-ds
----
+

> .. Change into the `charts/acm-simple-kmod-0.0.1` directory:

Author: Done

+
[source,terminal]
----
cd ..
----
.. Save the following YAML template for the chart as `Chart.yaml` in the `charts/acm-simple-kmod-0.0.1` directory:
+
[source,yaml]
----
apiVersion: v2
name: acm-simple-kmod
description: Build ACM enabled simple-kmod driver with SpecialResourceOperator
icon: https://avatars.githubusercontent.com/u/55542927
type: application
version: 0.0.1
appVersion: 1.6.4
----

> . From the `charts` directory, create the chart using the `helm package` command:

Reviewer: (Optional, suggesting because it says `helm package` immediately after and we don't usually put the command name in the step)
```suggestion
. From the `charts` directory, create the chart using the following command:
```

> . From the chart directory, create the chart using the helm package command:

Author: Done

+
[source,terminal]
----
$ helm package acm-simple-kmod-0.0.1/
----
+
.Example output
[source,terminal]
----
Successfully packaged chart and saved it to: <directory>/charts/acm-simple-kmod-0.0.1.tgz
----

> . Create a config map to store the chart files:

Reviewer: ```suggestion
. Create a config map to store the chart files.
```

> .. Create a directory for the config map files:

Reviewer: ```suggestion
.. Create a directory for the config map files by running the following command:
```

+
[source,terminal]
----
$ mkdir cm
----

> .. Copy the Helm chart into the `cm` directory:

Author: Done

+
[source, terminal]
----
$ cp acm-simple-kmod-0.0.1.tgz cm/acm-simple-kmod-0.0.1.tgz
----

> .. Create an index file specifying the Helm repo that contains the Helm chart:

Reviewer 2: ```suggestion
.. Create an index file specifying the Helm repository that contains the Helm chart:
```

Author: Done

> .. Create an index file specifying the Helm repository that contains the Helm chart:

Author: Done

+
[source, terminal]
----
$ helm repo index cm --url=cm://acm-simple-kmod/acm-simple-kmod-chart
----

> .. Create a namespace for the objects defined in the Helm chart:

Author: Done

+
[source, terminal]
----
$ oc create namespace acm-simple-kmod
----

> .. Create the config map object:

Author: Done

+
[source, terminal]
----
$ oc create cm acm-simple-kmod-chart --from-file=cm/index.yaml --from-file=cm/acm-simple-kmod-0.0.1.tgz -n acm-simple-kmod
----

> . Use the following `SpecialResourceModule` manifest to deploy the simple-kmod object using the Helm chart that you created in the config map. Save this YAML as `acm-simple-kmod.yaml`:

Reviewer 2: ```suggestion
. Use the following `SpecialResourceModule` manifest to deploy the `simple-kmod` object using the Helm chart that you created in the config map. Save this YAML file as `acm-simple-kmod.yaml`:
```

+
[source,yaml]
----
apiVersion: sro.openshift.io/v1beta1
kind: SpecialResourceModule
metadata:
    name: acm-simple-kmod
spec:
    namespace: acm-simple-kmod
    chart:
        name: acm-simple-kmod
        version: 0.0.1
        repository:
            name: acm-simple-kmod
            url: cm://acm-simple-kmod/acm-simple-kmod-chart
    set:
        kind: Values
        apiVersion: sro.openshift.io/v1beta1
        buildArgs:
            - name: "KMODVER"
              value: "SRO"
        registry: <your_registry>  <1>
        git:
            ref: master
            uri: https://github.com/openshift-psap/kvc-simple-kmod.git
    watch:
            - path: "$.metadata.labels.openshiftVersion"
              apiVersion: cluster.open-cluster-management.io/v1
              kind: ManagedCluster
              name: spoke1
----
<1> Specify the URL for a registry that you have configured.

> . Create the special resource module.

Reviewer: ```suggestion
. Create the special resource module by running the following command:
```

Author: Done

+
[source,terminal]
----
$ oc apply -f charts/examples/acm-simple-kmod.yaml
----

.Verification

> . Check the status of the build pods:

Author: Done

+
[source,terminal]
----
$ KUBECONFIG=~/hub/auth/kubeconfig oc get pod -n acm-simple-kmod
----
+
.Example Output
[source,terminal]
----
NAME                                                   READY   STATUS      RESTARTS   AGE
acm-simple-kmod-4-18-0-305-34-2-el8-4-x86-64-1-build   0/1     Completed   0          42m
----

> . Check that the policies have been created.

Reviewer 2: ```suggestion
. Check that the policies have been created by running the following command:
```

Author: Done

+
[source,terminal]
----
$ KUBECONFIG=~/hub/auth/kubeconfig oc get placementrules,placementbindings,policies -n acm-simple-kmod
----
+

> .Example Output

Reviewer 2: ```suggestion
.Example output
```

[source,terminal]
----
NAME                                                                      AGE   REPLICAS
placementrule.apps.open-cluster-management.io/acm-simple-kmod-placement   40m

NAME                                                                         AGE
placementbinding.policy.open-cluster-management.io/acm-simple-kmod-binding   40m

NAME                                                                 REMEDIATION ACTION   COMPLIANCE STATE   AGE
policy.policy.open-cluster-management.io/policy-acm-simple-kmod-ds   enforce              Compliant          40m
----

> . Check that the resources have been reconciled.

Reviewer 2: ```suggestion
. Check that the resources have been reconciled by running the following command:
```

Author: Done

+
[source,terminal]
----
$ KUBECONFIG=~/hub/auth/kubeconfig oc get specialresourcemodule acm-simple-kmod -o json | jq -r '.status'
----
+

> .Example Output

Reviewer 2: ```suggestion
.Example output
```

Author: Done

[source,terminal]
----
{
  "versions": {
    "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:6a3330ef5a178435721ff4efdde762261a9c55212e9b4534385e04037693fbe4": {
      "complete": true
    }
  }
}
----

> . Check that the resources are running in the spoke.

Reviewer 2: ```suggestion
. Check that the resources are running in the spoke by running the following command:
```

Author: Done

+
[source,terminal]
----
$ KUBECONFIG=~/spoke1/kubeconfig oc get ds,pod -n acm-simple-kmod
----
+

> Example Output

Reviewer 2: ```suggestion
Example output
```

Author: Done

----
+
[source,terminal]
----
$ cd charts/acm-simple-kmod-0.0.1/templates
----
. Create the buildconfig, policy, and placement.
.. Save this YAML template for the image stream and build config in the `templates` directory as `0001-buildconfig.yaml`.
+
[source,yaml]
----
apiVersion: build.openshift.io/v1
kind: BuildConfig
metadata:
    labels:
        app: {{ printf "%s-%s" .Values.specialResourceModule.metadata.name .Values.kernelFullVersion | replace "." "-" | replace "_" "-" | trunc 63 }}
    name: {{ printf "%s-%s" .Values.specialResourceModule.metadata.name .Values.kernelFullVersion | replace "." "-" | replace "_" "-" | trunc 63 }}
    annotations:
specialresource.openshift.io/wait: "true"
spec:
    nodeSelector:
        node-role.kubernetes.io/worker: ""
    runPolicy: "Serial"
    triggers:
        - type: "ConfigChange"
        - type: "ImageChange"
    source:
        git:
            ref: {{.Values.git.ref}}
            uri: {{.Values.git.uri}}
        type: Git
    strategy:
        dockerStrategy:
            dockerfilePath: Dockerfile.SRO
            buildArgs:
                - name: "IMAGE"
                value: {{ .Values.driverToolkitImage  }}

>                 {{- range $arg := .Values.buildArgs }}

Reviewer 3: indentation is off here, should be at the same level as "name".

Author: corrected

                - name: {{ $arg.name }}

>                 value: {{ $arg.value }}

Author: corrected

                {{- end }}
                - name: KVER

>                 value: {{ .Values.kernelFullVersion }}

Reviewer 3: indentation is off here, should be at the same level as "name".

Author: corrrected

----
.. Save this YAML template for the ACM policy in the `templates` directory as `0002-policy.yaml`.
... For a connected environment:
+
[source,yaml]
----
apiVersion: policy.open-cluster-management.io/v1
kind: Policy
metadata:
    name: policy-{{.Values.specialResourceModule.metadata.name}}-ds
    annotations:
        policy.open-cluster-management.io/categories: CM Configuration Management
        policy.open-cluster-management.io/controls: CM-2 Baseline Configuration
        policy.open-cluster-management.io/standards: NIST-CSF
spec:
    remediationAction: enforce
    disabled: false
    policy-templates:
        - objectDefinition:
            apiVersion: policy.open-cluster-management.io/v1
            kind: ConfigurationPolicy
            metadata:
                name: config-{{.Values.specialResourceModule.metadata.name}}-ds
            spec:
                remediationAction: enforce
                severity: low
                namespaceselector:
                    exclude:
                        - kube-*
                    include:
                        - '*'
                object-templates:
                    - complianceType: musthave

>                         objectDefinition:

Author: corrected

                        apiVersion: v1
                        kind: Namespace
                        metadata:
                            name: {{.Values.specialResourceModule.spec.namespace}}
                    - complianceType: mustonlyhave

>                         objectDefinition:

Reviewer 3: `objectDefinition` should be at the same level as `complianceType`

Author: corrected

                        apiVersion: v1
                        kind: ServiceAccount
                        metadata:
                            name: {{.Values.specialResourceModule.metadata.name}}
                            namespace: {{.Values.specialResourceModule.spec.namespace}}
                    - complianceType: mustonlyhave

>                         objectDefinition:

Reviewer 3: `objectDefinition` should be at the same level as `complianceType`

Author: corrected

Author: corrected

                            apiVersion: rbac.authorization.k8s.io/v1
                            kind: Role
                            metadata:
                                name: {{.Values.specialResourceModule.metadata.name}}
                                namespace: {{.Values.specialResourceModule.spec.namespace}}
                            rules:
                                - apiGroups:
                                    - security.openshift.io
                                resources:
                                    - securitycontextconstraints
                                verbs:
                                    - use
                                resourceNames:
                                    - privileged
                    - complianceType: mustonlyhave

>                         objectDefinition:

Reviewer 3: `objectDefinition` should be at the same level as `complianceType`

Author: corrected

                            apiVersion: rbac.authorization.k8s.io/v1
                            kind: RoleBinding
                            metadata:
                                name: {{.Values.specialResourceModule.metadata.name}}
                                namespace: {{.Values.specialResourceModule.spec.namespace}}
                            roleRef:
                                apiGroup: rbac.authorization.k8s.io
                                kind: Role
                                name: {{.Values.specialResourceModule.metadata.name}}
                            subjects:
                                - kind: ServiceAccount
                                name: {{.Values.specialResourceModule.metadata.name}}

>                                 namespace: {{.Values.specialResourceModule.spec.namespace}}

Reviewer 3: These should be at the same level as `kind`

Author: corrected

                    - complianceType: musthave

>                         objectDefinition:

Author: corrected

                                    spec:
                                        serviceAccount: {{.Values.specialResourceModule.metadata.name}}
                                        serviceAccountName: {{.Values.specialResourceModule.metadata.name}}
                                        containers:
                                            - image: {{.Values.registry}}/{{.Values.specialResourceModule.metadata.name}}-{{.Values.groupName.driverContainer}}:{{.Values.kernelFullVersion}}
                                                name: {{.Values.specialResourceModule.metadata.name}}
                                                imagePullPolicy: Always
                                                command: ["/sbin/init"]
                                                lifecycle:
                                                    preStop:
                                                        exec:
                                                            command: ["/bin/sh", "-c", "systemctl stop kmods-via-containers@{{.Values.specialResourceModule.metadata.name}}"]
                                                securityContext:

>                                                     privileged: true

Reviewer 3: These should be at the same level as `image`

Author: corrected

----
... For a disconnected environment:
+
[source,yaml]
----
apiVersion: policy.open-cluster-management.io/v1
kind: Policy
metadata:
    annotations:
        policy.open-cluster-management.io/categories: CM Configuration Management
        policy.open-cluster-management.io/controls: CM-2 Baseline Configuration
        policy.open-cluster-management.io/standards: NIST SP 800-53
    name: version-claim-policy
    namespace: default
spec:
    remediationAction: enforce
    disabled: false
    policy-templates:
        - objectDefinition:
            apiVersion: policy.open-cluster-management.io/v1
            kind: ConfigurationPolicy
            metadata:
                name: version-claim-policy-config
            spec:
                remediationAction: enforce
                severity: low
                namespaceselector:
                    exclude:
                        - kube-*
                    include:
                        - '*'
                object-templates:
                    - complianceType: musthave

>                         objectDefinition:

Author: corrected

                            apiVersion: cluster.open-cluster-management.io/v1alpha1
                            kind: ClusterClaim
                            metadata:
                                name: osimage.openshift.io
                            spec:
                                value: '{{ (lookup "config.openshift.io/v1" "ClusterVersion" "" "version").status.desired.image }}' <1>
----
<1> This ACM Policy creates the necessary cluster claim and keeps it in sync with changes if the spoke’s OCP version is upgraded.
.. Save this YAML template for the placement of policies in the `templates` directory as `0003-policy.yaml`.
... For a connected environment:
+
[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
    name: {{.Values.specialResourceModule.metadata.name}}-placement
spec:
    clusterConditions:
    - status: "True"

>         type: ManagedClusterConditionAvailable

Reviewer 3: Should be at the same level as `status`

Author: Corrected

    clusterSelector:
        matchExpressions:
            - key: name
                operator: NotIn
                values:

>                     - local-cluster

Reviewer 3: Should be at the same level as `key`

Author: corrected

---
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
    name: {{.Values.specialResourceModule.metadata.name}}-binding
placementRef:
    apiGroup: apps.open-cluster-management.io
    kind: PlacementRule
    name: {{.Values.specialResourceModule.metadata.name}}-placement
subjects:
    - apiGroup: policy.open-cluster-management.io
        kind: Policy

>         name: policy-{{.Values.specialResourceModule.metadata.name}}-ds

Reviewer 3: Should be at the same level as `apiGroup`

Author: corrected

----
... For a disconnected environment:
+
[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
    name: version-claim-placementrules
    namespace: default
spec:
    clusterSelector:
        matchExpressions:
            - key: name
                operator: NotIn
                values:

>                     - "local-cluster"

Reviewer 3: Should be at the same level as `key`

---
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
    name: version-claim-placementbinding
    namespace: default
placementRef:
    name: version-claim-placementrules
    kind: PlacementRule
    apiGroup: apps.open-cluster-management.io
subjects:
    - name: version-claim-policy
        kind: Policy

>         apiGroup: policy.open-cluster-management.io

Author: corrected

----
.. Create an index file specifying the Helm repo that contains the Helm chart:
+
[source, terminal]
----
$ helm repo index cm --url=cm://acm-simple-kmod/acm-simple-kmod-chart
----
.. Create a namespace for the objects defined in the Helm chart:
+
[source, terminal]
----
$ oc create namespace acm-simple-kmod
----
.. Create the config map object:
+
[source, terminal]
----
$ oc create cm acm-simple-kmod-chart --from-file=cm/index.yaml --from-file=cm/acm-simple-kmod-0.0.1.tgz -n acm-simple-kmod
----

. Use one of the following `SpecialResourceModule` manifests to deploy the simple-kmod object using the Helm chart that you created in the config map. Save this YAML as `acm-simple-kmod.yaml`:
.. For connected mode:
+
[source,yaml]
----
apiVersion: sro.openshift.io/v1beta1
kind: SpecialResourceModule
metadata:
    name: acm-simple-kmod
spec:
    namespace: acm-simple-kmod
    chart:
        name: acm-simple-kmod
        version: 0.0.1
        repository:
            name: acm-simple-kmod
            url: cm://acm-simple-kmod/acm-simple-kmod
        set:
            kind: Values
            apiVersion: sro.openshift.io/v1beta1
            buildArgs:
                - name: "KMODVER"

>                     value: "SRO"

Reviewer 3: Should be at the same level as `value`

Author: corrected

            registry: quay.io/openshift
            git:
                ref: master
                uri: https://github.com/openshift-psap/kvc-simple-kmod.git
        watch:
            - path: "$.metadata.labels.openshiftVersion"
                apiVersion: cluster.open-cluster-management.io/v1
                kind: ManagedCluster
                name: spoke1
----
.. For disconnected mode:
+
[source,yaml]
----
apiVersion: sro.openshift.io/v1beta1
kind: SpecialResourceModule
metadata:
    name: acm-simple-kmod
spec:
    namespace: acm-simple-kmod
    chart:
        name: acm-simple-kmod
        version: 0.0.1
        repository:
            name: acm-simple-kmod
            url: cm://acm-simple-kmod/acm-simple-kmod
        set:
            kind: Values
            apiVersion: sro.openshift.io/v1beta1
            buildArgs:
                - name: "KMODVER"

>                     value: "SRO"

Author: corrected

            registry: quay.io/openshift
            git:
                ref: master
                uri: https://github.com/openshift-psap/kvc-simple-kmod.git
        watch:
            - path: $.status.clusterClaims[?(@.name == 'osimage.openshift.io')].value"
                apiVersion: cluster.open-cluster-management.io/v1
                kind: ManagedCluster

>                 name: spoke1

Reviewer 3: Should be at the same level as `path`

Author: corrected

----
oc apply -f <file name>
----

.Procedure

. Create a templates directory and change into it.
+
[source,terminal]
----
$ mkdir -p charts/acm-simple-kmod-0.0.1/templates
----
+
[source,terminal]
----
$ cd charts/acm-simple-kmod-0.0.1/templates
----
. Create the buildconfig, policy, and placement.
.. Save this YAML template for the image stream and build config in the `templates` directory as `0001-buildconfig.yaml`.
+
[source,yaml]
----
apiVersion: build.openshift.io/v1
kind: BuildConfig
metadata:
    labels:
        app: {{ printf "%s-%s" .Values.specialResourceModule.metadata.name .Values.kernelFullVersion | replace "." "-" | replace "_" "-" | trunc 63 }}
    name: {{ printf "%s-%s" .Values.specialResourceModule.metadata.name .Values.kernelFullVersion | replace "." "-" | replace "_" "-" | trunc 63 }}
    annotations:
specialresource.openshift.io/wait: "true"
spec:
    nodeSelector:
        node-role.kubernetes.io/worker: ""
    runPolicy: "Serial"
    triggers:
        - type: "ConfigChange"
        - type: "ImageChange"
    source:
        git:
            ref: {{.Values.git.ref}}
            uri: {{.Values.git.uri}}
        type: Git
    strategy:
        dockerStrategy:

>             dockerfilePath: Dockerfile.SRO

Reviewer 3: This changed quite a bit. KVC support has been removed from the simple-kmod repo, so now we need these changes in the manifests: https://github.com/openshift/special-resource-operator/pull/229

Author: replaced the text per https://github.com/openshift/special-resource-operator/pull/229/files

                        - kube-*
                    include:
                        - '*'
                object-templates:
                    - complianceType: musthave
                      objectDefinition:
                        apiVersion: v1
                        kind: Namespace
                        metadata:
                            name: {{.Values.specialResourceModule.spec.namespace}}
                    - complianceType: mustonlyhave
                      objectDefinition:
                        apiVersion: v1
                        kind: ServiceAccount
                        metadata:
                            name: {{.Values.specialResourceModule.metadata.name}}
                            namespace: {{.Values.specialResourceModule.spec.namespace}}
                    - complianceType: mustonlyhave
                      objectDefinition:
                        apiVersion: rbac.authorization.k8s.io/v1
                        kind: Role
                        metadata:
                            name: {{.Values.specialResourceModule.metadata.name}}
                            namespace: {{.Values.specialResourceModule.spec.namespace}}

>                         rules:
>                         - apiGroups:
>                             - security.openshift.io
>                               resources:
>                             - securitycontextconstraints
>                             verbs:
>                             - use
>                             resourceNames:
>                             - privileged

Reviewer 3: ```suggestion
                        rules:
                        - apiGroups:
                            - security.openshift.io
                          resources:
                            - securitycontextconstraints
                          verbs:
                            - use
                          resourceNames:
                            - privileged
```

Author: Corrected

                                spec:
                                    serviceAccount: {{.Values.specialResourceModule.metadata.name}}
                                    serviceAccountName: {{.Values.specialResourceModule.metadata.name}}
                                    containers:
                                    - image: {{.Values.registry}}/{{.Values.specialResourceModule.metadata.name}}-{{.Values.groupName.driverContainer}}:{{.Values.kernelFullVersion}}
                                      name: {{.Values.specialResourceModule.metadata.name}}
                                      imagePullPolicy: Always
                                      command: ["/sbin/init"]
                                      lifecycle:
                                        preStop:
                                            exec:
                                                command: ["/bin/sh", "-c", "systemctl stop kmods-via-containers@{{.Values.specialResourceModule.metadata.name}}"]
                                            securityContext:

>                                                 privileged: true

Reviewer 3: This is also impacted by https://github.com/openshift/special-resource-operator/pull/229

----
.. Save this YAML template for the placement of policies in the `templates` directory as `0003-policy.yaml`.
+
[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
    name: {{.Values.specialResourceModule.metadata.name}}-placement
spec:
    clusterConditions:
    - status: "True"
      type: ManagedClusterConditionAvailable
    clusterSelector:
      matchExpressions:
      - key: name
        operator: NotIn
        values:
        - local-cluster
---
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
    name: {{.Values.specialResourceModule.metadata.name}}-binding
placementRef:
    apiGroup: apps.open-cluster-management.io
    kind: PlacementRule
    name: {{.Values.specialResourceModule.metadata.name}}-placement
subjects:
    - apiGroup: policy.open-cluster-management.io
      kind: Policy
      name: policy-{{.Values.specialResourceModule.metadata.name}}-ds
----
+
.. Change into the `charts/acm-simple-kmod-0.0.1` directory:
+
[source,terminal]
----
cd ..
----

> .. Save the following YAML for the chart as `Chart.yaml` in the `charts/acm-simple-kmod-0.0.1` directory:

Reviewer 3: Would a tree help here? Something like this:
```
$ tree
.
├── acm-simple-kmod.yaml
├── Chart.yaml
└── templates
    ├── 0001-buildconfig.yaml
    ├── 0002-policy.yaml
    └── 0003-placement.yaml

1 directory, 5 files
```

+
[source,yaml]
----
apiVersion: v2
name: acm-simple-kmod
description: Build ACM enabled simple-kmod driver with SpecialResourceOperator
icon: https://avatars.githubusercontent.com/u/55542927
type: application
version: 0.0.1
appVersion: 1.6.4
----

. Create a config map to store the chart files:
.. Create a directory for the config map files:
+
[source,terminal]
----
$ mkdir cm
----
.. Copy the Helm chart into the `cm` directory:
+
[source, terminal]
----

> $ cp acm-simple-kmod-0.0.1.tgz cm/acm-simple-kmod-0.0.1.tgz

Reviewer 3: I am missing the step to create the tgz, or did I skip it?

Reviewer 3: The command is `helm package <directory> -d <destination>`. In this case it would be `helm package charts/acm-simple-kmod-0.0.1 -d cm/`

Author: Done

----

. Check that the policies have been created.
+
[source,terminal]
----
KUBECONFIG=~/hub/auth/kubeconfig oc get placementrules,placementbindings,policies -n acm-simple-kmod
----
+
.Example Output
[source,terminal]
----
NAME                                                                      AGE   REPLICAS
placementrule.apps.open-cluster-management.io/acm-simple-kmod-placement   40m

NAME                                                                         AGE
placementbinding.policy.open-cluster-management.io/acm-simple-kmod-binding   40m

NAME                                                                 REMEDIATION ACTION   COMPLIANCE STATE   AGE
policy.policy.open-cluster-management.io/policy-acm-simple-kmod-ds   enforce              Compliant          40m
----

. Check that the resources have been reconciled.
+
[source,terminal]
----
KUBECONFIG=~/hub/auth/kubeconfig oc get specialresourcemodule acm-simple-kmod -o json | jq -r '.status'
----
+
.Example Output

[source,terminal]
----
{
  "versions": {
    "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:6a3330ef5a178435721ff4efdde762261a9c55212e9b4534385e04037693fbe4": {
      "complete": true
    }
  }
}

> ----

Reviewer 3: Does this command help from the spoke?
```
$ KUBECONFIG=~/spoke1/kubeconfig oc get ds,pod -n acm-simple-kmod
NAME                                                          DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
daemonset.apps/acm-simple-kmod-4-18-0-305-45-1-el8-4-x86-64   3         3         3       3            3           <none>          26m

NAME                                                     READY   STATUS    RESTARTS   AGE
pod/acm-simple-kmod-4-18-0-305-45-1-el8-4-x86-64-brw78   1/1     Running   0          26m
pod/acm-simple-kmod-4-18-0-305-45-1-el8-4-x86-64-fqh5h   1/1     Running   0          26m
pod/acm-simple-kmod-4-18-0-305-45-1-el8-4-x86-64-m9sfd   1/1     Running   0          26m
```

Reviewer 3: It shows the status after everything is running in the spoke.

version: 0.0.1
appVersion: 1.6.4
----

. Create a config map to store the chart files:
.. Create a directory for the config map files:
+
[source,terminal]
----
$ mkdir cm
----
.. Copy the Helm chart into the `cm` directory:
+
[source, terminal]
----
$ cp acm-simple-kmod-0.0.1.tgz cm/acm-simple-kmod-0.0.1.tgz
----
.. Create an index file specifying the Helm repo that contains the Helm chart:
+
[source, terminal]
----
$ helm repo index cm --url=cm://acm-simple-kmod/acm-simple-kmod-chart
----
.. Create a namespace for the objects defined in the Helm chart:
+
[source, terminal]
----
$ oc create namespace acm-simple-kmod
----
.. Create the config map object:
+
[source, terminal]
----
$ oc create cm acm-simple-kmod-chart --from-file=cm/index.yaml --from-file=cm/acm-simple-kmod-0.0.1.tgz -n acm-simple-kmod
----

. Use one of the following `SpecialResourceModule` manifests to deploy the simple-kmod object using the Helm chart that you created in the config map. Save this YAML as `acm-simple-kmod.yaml`:
.. For connected mode:
+
[source,yaml]
----
apiVersion: sro.openshift.io/v1beta1
kind: SpecialResourceModule
metadata:
    name: acm-simple-kmod
spec:
    namespace: acm-simple-kmod

>     chart:
>         name: acm-simple-kmod
>         version: 0.0.1
>         repository:
>             name: acm-simple-kmod
>             url: cm://acm-simple-kmod/acm-simple-kmod-chart
>         set:
>             kind: Values
>             apiVersion: sro.openshift.io/v1beta1
>             buildArgs:
>                 - name: "KMODVER"
>                   value: "SRO"
>             registry: <your registry>  <1>
>             git:
>                 ref: master
>                 uri: https://github.com/openshift-psap/kvc-simple-kmod.git
>         watch:
>             - path: "$.metadata.labels.openshiftVersion"
>               apiVersion: cluster.open-cluster-management.io/v1
>               kind: ManagedCluster
>               name: spoke1

Reviewer 3: ```suggestion
    chart:
        name: acm-simple-kmod
        version: 0.0.1
        repository:
            name: acm-simple-kmod
            url: cm://acm-simple-kmod/acm-simple-kmod-chart
    set:
        kind: Values
        apiVersion: sro.openshift.io/v1beta1
        buildArgs:
            - name: "KMODVER"
              value: "SRO"
        registry: <your registry>  <1>
        git:
            ref: master
            uri: https://github.com/openshift-psap/kvc-simple-kmod.git
    watch:
        - path: "$.metadata.labels.openshiftVersion"
          apiVersion: cluster.open-cluster-management.io/v1
          kind: ManagedCluster
          name: spoke1
```

Author: corrected in two instances
